autogpt_platform/autogpt_libs/autogpt_libs/utils/cache.py

```diff
@@ -138,48 +179,98 @@ async def async_wrapper(*args: P.args, **kwargs: P.kwargs):
                         if ttl_seconds is None:
+                            logger.debug(
+                                f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
+                            )
                             return cache_storage[key]
                         else:
                             cached_data = cache_storage[key]
                             if isinstance(cached_data, tuple):
                                 result, timestamp = cached_data
                                 if current_time - timestamp < ttl_seconds:
+                                    logger.debug(
+                                        f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
+                                    )
                                     return result
 
+                # Slow path: acquire lock for cache miss/expiry
+                async with _get_cache_lock():
+                    # Double-check: another coroutine might have populated cache
+                    if shared_cache:
+                        try:
+                            # Use GET directly (saves 1 round trip)
+                            cached_bytes = redis_client.get(redis_key)
+                            if cached_bytes is not None and isinstance(
+                                cached_bytes, bytes
+                            ):
+                                return pickle.loads(cached_bytes)
+                        except Exception as e:
+                            logger.error(
+                                f"Redis error during double-check for {target_func.__name__}: {e}"
+                            )
+                            # Continue to execute function
+                    else:
+                        if key in cache_storage:
+                            if ttl_seconds is None:
+                                return cache_storage[key]
+                            else:
+                                cached_data = cache_storage[key]
+                                if isinstance(cached_data, tuple):
+                                    result, timestamp = cached_data
+                                    if current_time - timestamp < ttl_seconds:
+                                        return result
+
                     # Cache miss - execute function
                     logger.debug(f"Cache miss for {target_func.__name__}")
                     result = await target_func(*args, **kwargs)
 
                     # Store result
-                    if ttl_seconds is None:
-                        cache_storage[key] = result
+                    if shared_cache:
+                        try:
+                            pickled_result = pickle.dumps(
+                                result, protocol=pickle.HIGHEST_PROTOCOL
+                            )
+                            if ttl_seconds is None:
+                                redis_client.set(redis_key, pickled_result)
```

Comment:
redis entry without expiry time is dangerous, can we remove this code path entirely ?

Reply:
I think we can remove the same functionality for in memory too for consistency, feel free to refactor this function so we can have a simpler implementation.